stan/math/prim/functor/multi_expression.hpp

```diff
@@ -0,0 +1,196 @@
+            std::enable_if_t<!Linear>* = nullptr>
+  void assign(const eigen_expressions_<T_expressions...>& expressions) {
+    constexpr bool is_first_row_major
+        = Eigen::internal::traits<std::decay_t<
+              std::tuple_element_t<0, std::tuple<T_expressions...>>>>::Flags
+          & Eigen::RowMajorBit;
+    size_t outer_dimension = is_first_row_major
+                                 ? std::get<0>(expressions.exprs_).rows()
+                                 : std::get<0>(expressions.exprs_).cols();
+    size_t inner_dimension = is_first_row_major
+                                 ? std::get<0>(expressions.exprs_).cols()
+                                 : std::get<0>(expressions.exprs_).rows();
+    for (size_t i = 0; i < outer_dimension; i++) {
+      for (size_t j = 0; j < inner_dimension; j++) {
+        index_apply<sizeof...(T_results)>([&](auto... Is) {
+          static_cast<void>(std::initializer_list<int>{
+              ((Eigen::internal::traits<std::decay_t<std::tuple_element_t<
+                            Is, std::tuple<T_expressions...>>>>::Flags
+                        & Eigen::RowMajorBit
+                    ? std::get<Is>(results_).coeffRef(j, i)
+                      = std::get<Is>(expressions.exprs_).coeff(j, i)
+                    : std::get<Is>(results_).coeffRef(i, j)
+                      = std::get<Is>(expressions.exprs_).coeff(i, j)),
+               0)...});
+        });
+      }
+    }
+  }
+
+ public:
+  /**
+   * Constructor.
+   * @param results results that will be calculated in same kernel
+   */
+  explicit eigen_results_(T_results&&... results)
+      : results_(std::forward<T_results>(results)...) {}
+
+  /**
+   * Assigning \c expressions_cl object to \c eigen_results_ object evals the
+   * expressions into results.
+   * @tparam T_expressions types of expressions
+   * @param expressions expressions
+   */
+  template <typename... T_expressions,
+            std::enable_if_t<sizeof...(T_results)
+                             == sizeof...(T_expressions)>* = nullptr>
+  void operator=(const eigen_expressions_<T_expressions...>& expressions) {
+    constexpr bool all_linear = std::min(
+        {static_cast<bool>(
+             Eigen::internal::traits<std::decay_t<T_expressions>>::Flags
+             & (Eigen::LinearAccessBit | Eigen::DirectAccessBit))...,
+         static_cast<bool>(
+             Eigen::internal::traits<std::decay_t<T_results>>::Flags
+             & (Eigen::LinearAccessBit | Eigen::DirectAccessBit))...});
+    constexpr int N_row_major = internal::static_sum(
+        static_cast<bool>(
+            Eigen::internal::traits<std::decay_t<T_expressions>>::Flags
+            & Eigen::RowMajorBit)...,
+        static_cast<bool>(
+            Eigen::internal::traits<std::decay_t<T_results>>::Flags
+            & Eigen::RowMajorBit)...);
+    constexpr int N_col_major
+        = sizeof...(T_results) + sizeof...(T_expressions) - N_row_major;
```

Comment:
Some of the traits are confusing Eigen internal things. Can you update the docs just detailing what each of these enums values we checks against means?

Reply:
Do we maintain some docs for Eigen internals? Or where would you like me to put that?

Reply:
I found the Eigen doc for these flags: https://eigen.tuxfamily.org/dox/group__flags.html